src/openrct2/core/Http.WinHttp.cpp

```diff
@@ -0,0 +1,223 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2020 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#if !defined(DISABLE_HTTP) && defined(_WIN32)
+
+#    include "Http.h"
+
+#    include "../Version.h"
+#    include "String.hpp"
+
+#    include <cstdio>
+#    include <stdexcept>
+#    define WIN32_LEAN_AND_MEAN
+#    include <windows.h>
+#    include <winhttp.h>
+
+namespace Http
+{
+    static constexpr char OPENRCT2_USER_AGENT[] = "OpenRCT2/" OPENRCT2_VERSION;
+
+    static void ThrowWin32Exception(const char* methodName)
+    {
+        auto errorCode = GetLastError();
+        auto msg = String::StdFormat("%s failed, error: %d.", methodName, errorCode);
+        throw std::runtime_error(msg);
+    }
+
+    static const wchar_t* GetVerb(Method method)
+    {
+        switch (method)
+        {
+            case Method::GET:
+                return L"GET";
+            case Method::POST:
+                return L"POST";
+            case Method::PUT:
+                return L"PUT";
+            default:
+                throw std::runtime_error("Unsupported verb.");
+        }
+    }
+
+    static std::wstring ReadHeader(HINTERNET hRequest, DWORD infoLevel, const wchar_t* name)
+    {
+        wchar_t headerBuffer[256]{};
+        auto headerBufferLen = (DWORD)std::size(headerBuffer);
+        if (!WinHttpQueryHeaders(hRequest, infoLevel, name, headerBuffer, &headerBufferLen, WINHTTP_NO_HEADER_INDEX))
+            ThrowWin32Exception("WinHttpQueryHeaders");
+        return std::wstring(headerBuffer, headerBufferLen);
+    }
+
+    static int32_t ReadStatusCode(HINTERNET hRequest)
+    {
+        auto wStatusCode = ReadHeader(hRequest, WINHTTP_QUERY_STATUS_CODE, L"StatusCode");
+        return std::stoi(wStatusCode);
+    }
+
+    static std::map<std::string, std::string> ReadHeaders(HINTERNET hRequest)
+    {
+        DWORD dwSize{};
+        WinHttpQueryHeaders(
+            hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, NULL, &dwSize, WINHTTP_NO_HEADER_INDEX);
+
+        std::wstring buffer;
+        buffer.resize(dwSize);
+        if (!WinHttpQueryHeaders(
+                hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, buffer.data(), &dwSize,
+                WINHTTP_NO_HEADER_INDEX))
+            ThrowWin32Exception("WinHttpQueryHeaders");
+
+        std::map<std::string, std::string> headers;
+        std::wstring wKey, wValue;
+        int state = 0;
+        int index = 0;
+        for (auto c : buffer)
```

Comment:
The whole code for iterating the headers seems overkill. Consider using WinHttpQueryHeaders correctly with the index parameter. You can easily iterate all headers by starting at index 0, after the call the variable will be set to the next index and if it runs out it will have ERROR_WINHTTP_HEADER_NOT_FOUND.

Reply:
The index parameter is only if you are dealing with multiple headers with the same key. Since we store them in a map, we can't store more than one anyway.
It also does not separate the key and value apart.

Reply:
Looks like you are right, sorry. I also wondered if we actually have to parse all headers, only the encoding is relevant so far.